File: frontend/src/types/generator.ts
export type ToolType = 'expdp' | 'exp' | 'impdp' | 'imp'
export type OracleVersion = '10g' | '11g' | '12c' | '19c' | '21c' | '23c'
export type AuthMethod = 'password' | 'wallet' | 'os'

export interface ConnectionParams {
  username: string
  password: string
  authMethod: AuthMethod
  connectType: 'service' | 'sid'
  serviceName: string
  sid: string
  host: string
  port: number
  sshEnabled: boolean
  sshHost: string
  sshPort: number
  sshUsername: string
  sshAuthMethod: 'password' | 'key'
  sshPassword: string
  sshKeyPath: string
  sshKeyPassphrase: string
  containerMode: '' | 'CDB' | 'PDB'
  pdbName: string
}

export interface ExpdpParams {
  directory: string
  dumpfile: string
  schemas: string[]
  tables: string[]
  tablespaces: string[]
  query: string
  content: 'ALL' | 'METADATA_ONLY' | 'DATA_ONLY'
  exclude: string[]
  include: string[]
  filesize: string
  reuseDumpfiles: boolean
  version: string
  flashbackScn: string
  flashbackTime: string
  encryption: string
  encryptionPassword: string
  encryptionMode: string
  encryptionColumnsOnly: string[]
  transportable: boolean
  transportFullCheck: boolean
  parallel: number
  logfile: string
  compression: string
  jobName: string
  estimate: string
}

export interface ExpParams {
  file: string
  owner: string[]
  tables: string[]
  query: string
  direct: boolean
  compress: boolean
  consistent: boolean
  statistics: string
  buffer: string
  rows: boolean
  indexes: boolean
  constraints: boolean
  grants: boolean
  triggers: boolean
  feedback: number
  fileSize: string
  resumable: boolean
  resumableName: string
  resumableTimeout: number
  recordlength: number
  objectConsistent: boolean
  parallel: number
  logfile: string
  compression: boolean
}

export interface ImpdpParams {
  directory: string
  dumpfile: string
  schemas: string[]
  tables: string[]
  remapSchema: string
  remapTablespace: string
  remapDatafile: string
  remapTable: string
  tableExistsAction: '' | 'SKIP' | 'APPEND' | 'TRUNCATE' | 'REPLACE'
  content: 'ALL' | 'METADATA_ONLY' | 'DATA_ONLY'
  exclude: string[]
  include: string[]
  transform: string
  sqlfile: string
  partitionOptions: string
  dataOptions: string
  disableArchiveLogging: boolean
  encryptionPassword: string
  version: string
  flashbackScn: string
  flashbackTime: string
  networkLink: string
  reuseDatafiles: boolean
  reuseDumpfiles: boolean
  parallel: number
  logfile: string
  compression: string
  jobName: string
  logtime: string
  metrics: boolean
}

export interface ImpParams {
  file: string
  fromuser: string[]
  touser: string[]
  tables: string[]
  commit: boolean
  ignore: boolean
  buffer: string
  statistics: string
  indexes: boolean
  constraints: boolean
  grants: boolean
  triggers: boolean
  rows: boolean
  feedback: number
  recordlength: number
  resumable: boolean
  resumableName: string
  resumableTimeout: number
  show: boolean
  destroy: boolean
  compile: boolean
  indexfile: string
  skipUnusableIndexes: boolean
  streamsize: string
  toidNovalidate: string[]
  filesize: string
  parallel: number
  logfile: string
  compression: boolean
}

export interface ValidationWarning {
  level: 'error' | 'warning' | 'info'
  field: string
  message: string
  suggestion?: string
}

export interface GenerateResponse {
  command: string
  parfile: string
  script: string
  script_bat: string
  warnings: ValidationWarning[]
  recommendations: string[]
  has_error: boolean
  directory_ddl: string
}

export interface GenerateRequest {
  tool: ToolType
  oracleVersion: OracleVersion
  connection: ConnectionParams
  params: Record<string, any>
}

File: backend/app/schemas/generator.py
from pydantic import BaseModel
from typing import Literal


class ConnectionParams(BaseModel):
    username: str = ""
    password: str = ""
    auth_method: Literal["password", "wallet", "os"] = "password"
    connect_type: Literal["service", "sid"] = "service"
    service_name: str = ""
    sid: str = ""
    host: str = "localhost"
    port: int = 1521
    # SSH Tunnel
    ssh_enabled: bool = False
    ssh_host: str = ""
    ssh_port: int = 22
    ssh_username: str = ""
    ssh_auth_method: Literal["password", "key"] = "password"
    ssh_password: str = ""
    ssh_key_path: str = ""
    ssh_key_passphrase: str = ""
    # Container Database
    container_mode: str = ""
    pdb_name: str = ""


class ExpdpParams(BaseModel):
    directory: str = ""
    dumpfile: str = ""
    schemas: list[str] = []
    tables: list[str] = []
    tablespaces: list[str] = []
    query: str = ""
    content: Literal["ALL", "METADATA_ONLY", "DATA_ONLY"] = "ALL"
    exclude: list[str] = []
    include: list[str] = []
    filesize: str = ""
    reuse_dumpfiles: bool = False
    version: str = ""
    flashback_scn: str = ""
    flashback_time: str = ""
    encryption: str = ""
    encryption_password: str = ""
    encryption_mode: str = ""
    encryption_columns_only: list[str] = []
    transportable: bool = False
    transport_full_check: bool = False
    parallel: int = 1
    logfile: str = ""
    compression: Literal["ALL", "DATA_ONLY", "METADATA_ONLY", "NONE"] = "NONE"
    job_name: str = ""
    estimate: Literal["BLOCKS", "STATISTICS", ""] = ""


class ExpParams(BaseModel):
    file: str = ""
    owner: list[str] = []
    tables: list[str] = []
    query: str = ""
    direct: bool = False
    compress: bool = False
    consistent: bool = False
    statistics: str = "ESTIMATE"
    buffer: str = ""
    rows: bool = True
    indexes: bool = True
    constraints: bool = True
    grants: bool = True
    triggers: bool = True
    feedback: int = 0
    file_size: str = ""
    resumable: bool = False
    resumable_name: str = ""
    resumable_timeout: int = 0
    recordlength: int = 0
    object_consistent: bool = False
    parallel: int = 1
    logfile: str = ""
    compression: bool = False


class ImpdpParams(BaseModel):
    directory: str = ""
    dumpfile: str = ""
    schemas: list[str] = []
    tables: list[str] = []
    remap_schema: str = ""
    remap_tablespace: str = ""
    remap_datafile: str = ""
    remap_table: str = ""
    table_exists_action: Literal["SKIP", "APPEND", "TRUNCATE", "REPLACE"] = ""
    content: Literal["ALL", "METADATA_ONLY", "DATA_ONLY"] = "ALL"
    exclude: list[str] = []
    include: list[str] = []
    transform: str = ""
    sqlfile: str = ""
    partition_options: str = ""
    data_options: str = ""
    disable_archive_logging: bool = False
    encryption_password: str = ""
    version: str = ""
    flashback_scn: str = ""
    flashback_time: str = ""
    network_link: str = ""
    reuse_datafiles: bool = False
    reuse_dumpfiles: bool = False
    parallel: int = 1
    logfile: str = ""
    compression: Literal["ALL", "DATA_ONLY", "METADATA_ONLY", "NONE"] = "NONE"
    job_name: str = ""
    logtime: Literal["ALL", "HEADER", "LOGFILE", "STATUS", ""] = ""
    metrics: bool = False


class ImpParams(BaseModel):
    file: str = ""
    fromuser: list[str] = []
    touser: list[str] = []
    tables: list[str] = []
    commit: bool = False
    ignore: bool = False
    buffer: str = ""
    statistics: str = "RECALCULATE"
    indexes: bool = True
    constraints: bool = True
    grants: bool = True
    triggers: bool = True
    rows: bool = True
    feedback: int = 0
    recordlength: int = 0
    resumable: bool = False
    resumable_name: str = ""
    resumable_timeout: int = 0
    show: bool = False
    destroy: bool = False
    compile: bool = False
    indexfile: str = ""
    skip_unusable_indexes: bool = False
    streamsize: str = ""
    toid_novalidate: list[str] = []
    filesize: str = ""
    parallel: int = 1
    logfile: str = ""
    compression: bool = False


class GenerateRequest(BaseModel):
    tool: Literal["expdp", "exp", "impdp", "imp"]
    oracle_version: Literal["10g", "11g", "12c", "19c", "21c", "23c"] = "19c"
    connection: ConnectionParams
    params: dict


class ValidationWarning(BaseModel):
    level: Literal["error", "warning", "info"]
    field: str
    message: str
    suggestion: str | None = None


class GenerateResponse(BaseModel):
    command: str
    parfile: str
    script: str
    warnings: list[ValidationWarning]
    recommendations: list[str]
    directory_ddl: str = ""
    steps: list[str] = []
